File: tests/test_support_contract.cpp
#include "mfem.hpp"

#include "../src/ofdg.hpp"

#include <iostream>
#include <string>

using namespace mfem;

namespace
{

void ConstructFilter(FiniteElementSpace &space)
{
   ofdg::OFDG filter(&space, BasisType::GaussLegendre);
}

} // namespace

int main(int argc, char *argv[])
{
   MFEM_VERIFY(argc >= 2, "Expected a support-contract test mode.");
   const std::string mode = argv[1];

   if (mode == "supported")
   {
      Mesh mesh = Mesh::MakeCartesian2D(
         2, 2, Element::QUADRILATERAL, true, 1.0, 1.0);
      DG_FECollection collection(2, 2);
      FiniteElementSpace space(&mesh, &collection);
      ConstructFilter(space);
      return 0;
   }

   if (mode == "mixed" || mode == "curved")
   {
      MFEM_VERIFY(argc == 3, "Expected the MFEM mesh path.");
      Mesh mesh(argv[2], 1, 1);
      DG_FECollection collection(2, mesh.Dimension());
      FiniteElementSpace space(&mesh, &collection);
      ConstructFilter(space);
      return 0;
   }

   if (mode == "variable-order")
   {
      Mesh mesh = Mesh::MakeCartesian1D(3, 1.0);
      DG_FECollection collection(2, 1);
      FiniteElementSpace space(&mesh, &collection);
      space.SetElementOrder(1, 1);
      ConstructFilter(space);
      return 0;
   }

   if (mode == "modified")
   {
      Mesh mesh = Mesh::MakeCartesian1D(3, 1.0);
      DG_FECollection collection(2, 1);
      FiniteElementSpace space(&mesh, &collection);
      ofdg::OFDG filter(&space, BasisType::GaussLegendre);
      GridFunction state(&space);
      state = 1.0;
      mesh.UniformRefinement();
      Vector means;
      filter.ComputeMean(state, means);
      return 0;
   }

   std::cerr << "Unknown support-contract mode: " << mode << '\n';
   return 2;
}

File: tests/test_support_contract.py
#!/usr/bin/env python3
"""Process-level checks for contracts that deliberately terminate via MFEM."""

from __future__ import annotations

import argparse
import subprocess
from pathlib import Path


def run(binary: Path, mode: str, mesh: Path | None = None) -> subprocess.CompletedProcess[str]:
    command = [str(binary), mode]
    if mesh is not None:
        command.append(str(mesh))
    return subprocess.run(command, capture_output=True, text=True, timeout=20, check=False)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("binary", type=Path)
    parser.add_argument("--mfem-data", type=Path, required=True)
    arguments = parser.parse_args()

    supported = run(arguments.binary, "supported")
    if supported.returncode != 0:
        raise RuntimeError(f"supported mesh was rejected:\n{supported.stderr}")

    rejected = [
        ("mixed", arguments.mfem_data / "square-mixed.mesh", "multiple element geometries"),
        ("curved", arguments.mfem_data / "disc-nurbs.mesh", "curvilinear or non-affine"),
        ("variable-order", None, "variable polynomial order"),
        ("modified", None, "changed after construction"),
    ]
    for mode, mesh, expected in rejected:
        result = run(arguments.binary, mode, mesh)
        output = result.stdout + result.stderr
        if result.returncode == 0:
            raise RuntimeError(f"unsupported {mode} case unexpectedly succeeded")
        if expected not in output:
            raise RuntimeError(
                f"unsupported {mode} case did not report {expected!r}:\n{output}"
            )

    print("Supported-space contract tests passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
